### PR TITLE
Fix Microsoft.ReactNative unit tests

### PR DESCRIPTION
**Pull request by vmoroz**
The PR fixes the following issues:
- Failing new JSI test for the error handling - fix in Chakra JSI and ABI JSI implementations
- Managed code generation tests that required update to WinRT package versions - thank you @dannyvv for the help with it!

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7984)

### CODE REVIEW

**Review by tudorms — APPROVED** [review]

File: vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs [hunk 1]
```diff
@@ -110,8 +110,8 @@ protected async Task<(CodeAnalyzer, INamedTypeSymbol)> AnalyzeFileAsync(string c
         $@"{win10SdkFolder}\References\{win10SdkVersion}\Windows.Devices.Printers.PrintersContract\1.0.0.0\Windows.Devices.Printers.PrintersContract.winmd",
         $@"{win10SdkFolder}\References\{win10SdkVersion}\Windows.Devices.SmartCards.SmartCardBackgroundTriggerContract\3.0.0.0\Windows.Devices.SmartCards.SmartCardBackgroundTriggerContract.winmd",
         $@"{win10SdkFolder}\References\{win10SdkVersion}\Windows.Devices.SmartCards.SmartCardEmulatorContract\6.0.0.0\Windows.Devices.SmartCards.SmartCardEmulatorContract.winmd",
-        $@"{win10SdkFolder}\References\{win10SdkVersion}\Windows.Foundation.FoundationContract\3.0.0.0\Windows.Foundation.FoundationContract.winmd",
-        $@"{win10SdkFolder}\References\{win10SdkVersion}\Windows.Foundation.UniversalApiContract\8.0.0.0\Windows.Foundation.UniversalApiContract.winmd",
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs:114]: these versions are fragile (they can be updated, and it makes it harder to update the target SDK version). Can the test be updated to discover which contract version is available in a given SDK?

File: vnext/Microsoft.ReactNative.Cxx/JSI/JsiAbiApi.cpp [hunk 2]
```diff
@@ -634,9 +634,23 @@ bool JsiAbiRuntime::instanceOf(const Object &o, const Function &f) try {
...
+  AutoRestore(T *var, T value) : m_var{var}, m_value{std::exchange(*var, value)} {}
+
+  ~AutoRestore() {
+    *m_var = m_value;
+  }
+
+ private:
+  T *m_var;
+  T m_value;
+};
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/JSI/JsiAbiApi.cpp:648]: consider putting this in a common header if you're going to use it from more than one place

**Review by asklar — APPROVED** [review]

File: vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs (hunk 1 above)
Comment (dannyvv) [reply vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs:114]: Filed: #7985

Comment (dannyvv) [reply vnext/Microsoft.ReactNative.Managed.CodeGen.UnitTests/AnalysisTestBase.cs:114]: Filed: #7985

**Pull request merged**